### PR TITLE
feat: add optional name for activity descriptions

### PR DESCRIPTION
**Pull request by Anyitechs**
This PR closes issue #127  and the implementation follows the conversation on the issue thread.

### CODE REVIEW

File: sim-lib/src/lib.rs [hunk 1]
```diff
@@ -573,19 +578,25 @@ impl Simulation {
...
+            for (index, description) in self.activity.iter().enumerate() {
                 let activity_generator = DefinedPaymentActivity::new(
                     description.destination.clone(),
                     Duration::from_secs(description.interval_secs.into()),
                     description.amount_msat,
                 );
 
+                let activity_name = match &description.activity_name {
+                    Some(name) => name.clone(),
+                    None => index.to_string(),
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:590]: perhaps "index {}" so that users understand where the number comes from

File: sim-lib/src/lib.rs [hunk 2]
```diff
@@ -790,13 +803,14 @@ async fn consume_events(
...
     source: NodeInfo,
+    activity_name: String,
     network_generator: Arc<Mutex<N>>,
     node_generator: Box<A>,
     sender: Sender<SimulationEvent>,
     shutdown: Trigger,
     listener: Listener,
 ) {
-    log::info!("Started activity producer for {source}: {node_generator}.");
+    log::info!("Started {activity_name} => activity producer for {source}: {node_generator}.");
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:813]: Let's have the `activity_name` as a prefix for all te logs in `produce_events` so that they're easily reference-able. Extra points if it's propagated through to other activity-relevant logging.

File: sim-lib/src/lib.rs [hunk 3]
```diff
@@ -637,6 +648,7 @@ impl Simulation {
             generators.push(ExecutorKit {
                 source_info: node_info.clone(),
                 network_generator: network_generator.clone(),
+                activity_name: "Random activity".to_string(),
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:651]: This doesn't really add much, I'd just have no name for randomly generated activity.

**Review by carlaKC — COMMENTED** [review]
Conceptually looks good, a few comments. Feel free to pull out of draft once addressed.

File: sim-lib/src/lib.rs (hunk 1 above)
Comment (Anyitechs) [reply sim-lib/src/lib.rs:590]: Sure! I'll fix that alongside the other changes and update the PR.

**Comment by Anyitechs** [comment]
@carlaKC  I made the requested changes. Let me know if this looks good now. Thank you.

**Review by enigbe — CHANGES_REQUESTED** [review]
@Anyitechs Good job on the PR. Here are a few comments I have:

1. To make the new `activity_name` field even clearer for users, I'd recommend adding some documentation to the README. This could explain its optional nature and how it's handled in the code.

2. I noticed that the `activity_name` or `index` is used in logs related to activities and at most the simulation events (when they're sent to the events consumer). I'm curious about why we'd have a name/index prefix that doesn't extend to the simulation outputs and results as well. Shouldn't there be consistency across board? I don't think it will be easy to relate `activity -> event -> output -> result` logs if only the activities and the events have the prefix. I know this is out of the scope of the issue but happy to get your (as well as @carlaKC's) thoughts on this.

**Comment by Anyitechs** [comment]
> @Anyitechs Good job on the PR. Here are a few comments I have:

Thank you.

> 
> 1. To make the new `activity_name` field even clearer for users, I'd recommend adding some documentation to the README. This could explain its optional nature and how it's handled in the code.

Sure! I'll update the README in a seperate PR following this.

> 2. I noticed that the `activity_name` or `index` is used in logs related to activities and at most the simulation events (when they're sent to the events consumer). I'm curious about why we'd have a name/index prefix that doesn't extend to the simulation outputs and results as well. Shouldn't there be consistency across board? I don't think it will be easy to relate `activity -> event -> output -> result` logs if only the activities and the events have the prefix. I know this is out of the scope of the issue but happy to get your (as well as @carlaKC's) thoughts on this.

I agree @enigbe . I guess that's what @carlaKC referred to as "activity-related" logging above. I'll work on that and update the PR.

**Comment by carlaKC** [comment]
> Sure! I'll update the README in a seperate PR following this.

Please update readme in the same PR, it's easy for things to get out of sync/be forgotten otherwise.

**Review by carlaKC — COMMENTED** [review]
Shaping up! 

I think that this still needs better handling for random activity - what we have right now will print odd looking lines.

File: README.md [hunk 4]
```diff
@@ -136,8 +136,11 @@ of the random activity that is generated:
 ### Setup - Defined Activity
 If you would like SimLN to generate a specific payments between source 
 and destination nodes, you can provide `activity` descriptions of the 
-source, destination, frequency and amount for payments that you'd like 
-to execute. Note that `source` nodes *must* be contained in `nodes`, 
+source, destination, frequency, amount for payments that you'd like 
+to execute and an optional name for each `activity` description that is used
+as the logging prefix to relate logs to their corresponding activity description 
+(The index of each `activity` description will be used if a name is not 
+provided for the `activity` description). Note that `source` nodes *must* be contained in `nodes`,
```
[the remark this answers is not in the document]
Comment (Anyitechs) [reply README.md:143]: Sure, I'll update that. Thank you.

File: sim-lib/src/lib.rs [hunk 5]
```diff
@@ -819,32 +838,32 @@ async fn produce_events<N: DestinationGenerator + ?Sized, A: PaymentGenerator +
                 let amount = match node_generator.payment_amount(capacity) {
                     Ok(amt) => {
                         if amt == 0 {
-                            log::debug!("Skipping zero amount payment for {source} -> {destination}.");
+                            log::debug!("{activity_name} activity => Skipping zero amount payment for {source} -> {destination}.");
```
[the remark this answers is not in the document]
Comment (Anyitechs) [reply sim-lib/src/lib.rs:841]: Ok. I'll replace that. 

And if a name isn't set, it should print the `index` of the `activity`. But I'll confirm this is true for this as well.

File: sim-cli/src/main.rs [hunk 6]
```diff
@@ -187,6 +187,7 @@ async fn main() -> anyhow::Result<()> {
             destination,
             interval_secs: act.interval_secs,
             amount_msat: act.amount_msat,
+            activity_name: act.activity_name,
```
[the remark this answers is not in the document]
Comment (Anyitechs) [reply sim-cli/src/main.rs:190]: Yes, I agree. How best do you think we can make them unique? Does appending the event name to the activity name work?

**Review by carlaKC — COMMENTED** [review]
Needs a rebase on master!

File: sim-cli/src/main.rs [hunk 7]
```diff
@@ -182,11 +182,20 @@ async fn main() -> anyhow::Result<()> {
             },
         };
 
+        for validated_activity in &validated_activities {
+            if validated_activity.activity_name == act.activity_name {
+                anyhow::bail!(LightningError::ValidationError(
+                    "Duplicate activity name is not allowed".to_string(),
```
Comment (carlaKC) [inline sim-cli/src/main.rs:188]: It would be helpful to provide the user with the name so they know what to fix?

File: sim-lib/src/lib.rs [hunk 8]
```diff
@@ -935,19 +960,32 @@ impl PaymentResultLogger {
...
-            "Processed {} payments sending {} msat total with {:.2}% success rate.",
-            total_payments,
-            self.total_sent,
-            (self.success_payment as f64 / total_payments as f64) * 100.0
-        )
+
+        if self.activity_name.is_empty() {
+            write!(
+                f,
+                "Processed {} payments sending {} msat total with {:.2}% success rate.",
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:974]: Surely we can do this with fewer LOC?

File: sim-cli/src/main.rs [hunk 9]
```diff
@@ -182,11 +182,20 @@ async fn main() -> anyhow::Result<()> {
             },
         };
 
+        for validated_activity in &validated_activities {
```
Comment (carlaKC) [inline sim-cli/src/main.rs:185]: Build a hashmap like we do for alias and pubkey rather than iterating through every activity every time

File: README.md [hunk 10]
```diff
@@ -84,17 +92,18 @@ Whereas for CLN nodes, the following information is required:
...
+- Defined activity: provide descriptions of specific payments that
   you would like the generator to execute.
 
 ### Setup - Random Activity
 
-To run the simulator with random activity generation, you just need to 
-provide a set of nodes and the simulator will generate activity based 
-on the topology of the underlying graph. Note that payments will only 
-be sent between the `nodes` that are provided so that liquidity does 
+To run the simulator with random activity generation, you just need to
```
Comment (carlaKC) [inline README.md:103]: Please pull wholesale formatting changes out into their own commit

**Comment by Anyitechs** [comment]
Thank you for the detailed review @carlaKC. I've addressed all the feedbacks, rebased, and updated the PR as well.

Let me know if there's anything else I need to look at.

File: sim-lib/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (carlaKC) [reply sim-lib/src/lib.rs]: We don't set an `activity_name` for randomly generated activity (above, line 776) right?

File: sim-lib/src/lib.rs [hunk 11]
```diff
@@ -1127,12 +1122,17 @@ impl PaymentResultLogger {
     }
 }
 
+// TODO: Add activity name and handle the initial processing log with few LOC
 impl Display for PaymentResultLogger {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         let total_payments = self.success_payment + self.failed_payment;
+
+        // if self.activity_name
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:1130]: Incomplete comment and TODO left in this commit?

**Review by carlaKC — COMMENTED** [review]
Looks like the functionality is mostly here (though I think we still need to clean up the logging for random activity). 

On a higher level, the commit structure in this PR needs some work. Most of the changes can be squashed down to a single commit, and formatting changes / docs / unrelated fixups left in their own commit.

File: sim-lib/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Anyitechs) [reply sim-lib/src/lib.rs]: Yes, we don't set an `activity_name` for randomly generated activity. I see your point on this logging ` activity` if a name isn't set for it.

How best do you think we can handle this? I had a name set for it before but you earlier mentioned we should take it off.

[the remark this answers is not in the document]
Comment (carlaKC) [reply sim-lib/src/lib.rs]: By only logging the prefix if we've got a value there?

`if self.activity_name != "" {  format!("{} activity:", self.activity_name) } else {"".to_string()}`

Could inline, pulled out into a impl of `Payment`, or leave `activity_name` as an option so that we can use some of the `uwrap_or` functionality.

[the remark this answers is not in the document]
Comment (Anyitechs) [reply sim-lib/src/lib.rs]: Thank you for the clarification. I've addressed this and updated the PR.

**Comment by Anyitechs** [comment]
> Looks like the functionality is mostly here (though I think we still need to clean up the logging for random activity).
> 
> On a higher level, the commit structure in this PR needs some work. Most of the changes can be squashed down to a single commit, and formatting changes / docs / unrelated fixups left in their own commit.

Thank you. I've addressed the feedbacks and neatened up the commit structure.

File: README.md (hunk 10 above)
Comment (carlaKC) [reply README.md:103]: This still hasn't been addressed?

If you want to run all these formatting changes on the README, please run them first and then make your changes.

File: sim-lib/src/lib.rs (hunk 11 above)
Comment (carlaKC) [reply sim-lib/src/lib.rs]: Also unaddressed?

File: README.md
[the remark this answers is not in the document]
Comment (carlaKC) [reply README.md:1]: WDYM? This commit also has all of the rust changes in it?

File: sim-lib/src/lib.rs [hunk 12]
```diff
@@ -358,6 +358,16 @@ struct Payment {
     dispatch_time: SystemTime,
 }
 
+impl Payment {
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:361]: Can be squashed with the original rust changes?

File: sim-lib/src/lib.rs (hunk 11 above)
Comment (carlaKC) [reply sim-lib/src/lib.rs]: Then the [fixup should be squashed](https://jordanelver.co.uk/blog/2020/06/04/fixing-commits-with-git-commit-fixup-and-git-rebase-autosquash/) as requested.

File: README.md
[the remark this answers is not in the document]
Comment (Anyitechs) [reply README.md:1]: I'm sorry I was referring to this [commit](https://github.com/bitcoin-dev-project/sim-ln/pull/170/commits/19f7162aaa229e15185fa4de8aff1c5810987d1f) and I don't think it has all of the rust changes in it. But I'll address the format issue in the first commit.

[the remark this answers is not in the document]
Comment (carlaKC) [reply README.md:1]: Please see [suggested structure](https://github.com/bitcoin-dev-project/sim-ln/pull/170#issuecomment-2073242761) and [contribution guidelines](https://github.com/bitcoin-dev-project/sim-ln/blob/main/CONTRIBUTING.md#commit-history).

File: sim-lib/src/lib.rs (hunk 11 above)
Comment (Anyitechs) [reply sim-lib/src/lib.rs]: Ok. Thank you.

**Comment by Anyitechs** [comment]
> > Thank you. I've addressed the feedbacks and neatened up the commit structure.
> 
> Have you? The first commit is still a mix of formatting and logical changes, and has a dangling TODO comment.
> 
> To me, the logical commit structure would be:
> 
> * Formatting changes to the README (or remove all of those)
> * Commit with actual rust changes
> * Commit with documentation updates for features
> 
> Right now everything is intermingled in [the first commit](https://github.com/bitcoin-dev-project/sim-ln/pull/170/commits/1f2920772a2e33e8215c3a13f447fe49a82586c0) - formatting changes to the readme, and rust code.

Thank you @carlaKC for taking the time to review this and providing feedbacks as well. I really appreciate it.

I've worked on the commit structure and currently squashed it into two commits:

- The actual code implementation, and
- The documentation update of the feature

Let me know if I missed anything. Thank you.

**Review by carlaKC — CHANGES_REQUESTED** [review]
Thanks for updating the commit ordering! 

Please try to test changes with a few standard inputs before you push/request review - there's a panic in this change. 

The various logging issues that I've pointed out also become more apparent when you spend some time running the project and getting familiar. I know that these seem like frustrating nits, but when we're looking at a change like this which really is just aesthetic we really do need to get things right to justify changing the code at all.

**Comment by carlaKC** [comment]
If it's useful to getting this in, we can also just re-add `random activity` as a name, but perhaps with an index. I don't like it but it's not really I hill I'm willing to die on - hopefully that simplifies all the logging.

**Comment by carlaKC** [comment]
Needs rebase!

**Comment by Anyitechs** [comment]
> Ping @Anyitechs any updated here?

Yes, will update the PR by EOD.

**Comment by Anyitechs** [comment]
> > Ping @Anyitechs any updated here?
> 
> Yes, will update the PR by EOD.

I just updated the PR now.

File: sim-lib/src/lib.rs [hunk 13]
```diff
@@ -1116,30 +1150,40 @@ struct PaymentResultLogger {
...
-            "Processed {} payments sending {} msat total with {:.2}% success rate.",
-            total_payments,
-            self.total_sent,
-            (self.success_payment as f64 / total_payments as f64) * 100.0
-        )
+        for (activity_name, activity_result) in &self.activity_results {
+            let total_payments = activity_result.success_payment + activity_result.failed_payment;
+            writeln!(
+                f,
+                "{} Processed {} payments sending {} msat total with {:.2}% success rate.",
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:1179]: The intent behind [this comment](https://github.com/bitcoin-dev-project/sim-ln/pull/170#discussion_r1578293858) was to remove the names from `PaymentResultLogger` completely - we don't want to print these results on a per-activity basis, it'll be too spammy where the whole point here is to give a very high level aggregate.

I believe we can also remove `activity_name` from `Payment` completely if we don't have this logging?

File: sim-lib/src/lib.rs [hunk 14]
```diff
@@ -352,6 +358,16 @@ struct Payment {
     dispatch_time: SystemTime,
 }
 
+impl Payment {
+    fn activity_prefix(&self) -> String {
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:362]: nit: add a comment explaining usage here / that this should be added before a log with no space

File: sim-lib/src/lib.rs [hunk 15]
```diff
@@ -352,6 +358,16 @@ struct Payment {
...
+        if !self.activity_name.is_empty() {
+            format!("{} activity:", self.activity_name)
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:364]: nit: should have a space after `:` otherwise it reads `X activity:Payment` when we use it below

File: sim-lib/src/lib.rs [hunk 16]
```diff
@@ -788,10 +812,11 @@ impl Simulation {
             network_generator.lock().await
         );
 
-        for (node_info, capacity) in active_nodes.values() {
+        for (index, (node_info, capacity)) in active_nodes.values().enumerate() {
             generators.push(ExecutorKit {
                 source_info: node_info.clone(),
                 network_generator: network_generator.clone(),
+                activity_name: format!("Index {index} random"),
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:819]: nit: `Random index {}`

File: sim-lib/src/lib.rs [hunk 17]
```diff
@@ -788,10 +812,11 @@ impl Simulation {
             network_generator.lock().await
         );
 
-        for (node_info, capacity) in active_nodes.values() {
+        for (index, (node_info, capacity)) in active_nodes.values().enumerate() {
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:815]: nit: just use i, more idiomatic

**Review by carlaKC — CHANGES_REQUESTED** [review]
Can be rebased on master to fix unrelated build failure. 

I think that this change is high impact enough if we just have logging in `produce_events`, so final comment is to remove this entirely from `PaymentResultLogger`.

**Review by carlaKC — CHANGES_REQUESTED** [review]
There are some superfluous changes leftover from previous iterations of this PR, and `activity_name` isn't being used in logs in `produce_events` in several places.

**Comment by carlaKC** [comment]
Closing due to inactivity.
